File: cogs/moderation/bans.js
// eslint-disable-next-line no-unused-vars
const Discord = require('discord.js');
require('../throwError.js')();
var bot = {};

var ban = function (msg) {
    let args = msg.content.trim().split(' ');
    if (args.length === 0) return msg.reply('Musisz podać użytkownika! \n> `!ban <id/@> (0-7 dni) (Powód)` \n> Dni określają czas usunięcia wiadomości użytkownika.');
    let member = msg.mentions.members.first();
    let userid;
    if (member) userid = member.id;
    if (!member) {
        if (args[0].startsWith('<@') && args[0].endsWith('>')) {
            const id = args[0].slice(2, -1);
            member = msg.guild.members.fetch(`${id}`)
            .then(userid = id)
            // eslint-disable-next-line no-undef
            .catch(error => throwError(msg, error, "18", "ban " + msg.content));
            //userid = `${id}`;
        }
        else {
            member = msg.guild.members.fetch(`${args[0]}`)
            .then(userid = args[0])
            // eslint-disable-next-line no-undef
            .catch(error => throwError(msg, error, "25", "ban " + msg.content));
            //userid = `${args[0]}`;
        }
    }
    args.shift();
    let days = args[0];
    let daysMsg = "";
    days === parseInt(args[0])
    if (days <= 7 && days >= 0) {
        daysMsg = `Dodatkowo usunięto wiadomości z ostatnich ${days} dni.`
        args.shift();
    }
    else days = 0;
    let reason = args.join(' ');
    let silentMsg = "";
    // if (!member.bannable) {
    //     msg.reply("Nie mogę zbanować tego użytkownika!")
    //     return
    // }
    if ((member == msg.author) || (msg.author.id == userid)) {
        msg.reply("Dlaczego chcesz zbanować samego siebie?")
        return
    }
    if (!reason) {
        reason = "Brak powodu"
    }
    if (reason.includes("-s")) {
        silentMsg = " [Flaga -s]"
    }
    msg.guild.bans.create(userid, {days: days, reason: reason})
        .then(msg.reply(`Zrozumiano, użytkownik <@${userid}> został zbanowany za ${reason}.${silentMsg} ${daysMsg}`))
        .catch(console.error);
    //msg.reply(`Zrozumiano, użytkownik <@${userid}> został zbanowany za ${reason}.${silentMsg} ${daysMsg}`);
}

var banold = function (msg) {
    if (!msg.member.permissions.has("BAN_MEMBERS")) {
        msg.reply("Brak uprawnień!");
        return
    }
    //const args = msg.content.trim().split(' ');
    const args = msg.content.trim().replace(/  +/g, ' ').replace(/<@.?[0-9]*?> /g, '');
    const user = msg.mentions.members.first() || msg.guild.members.cache.get(args[0]) || msg.guild.members.cache.find(x => x.user.username.toLowerCase() === args.slice(0).join(' ') || x.user.username === args[0]);
    //const user = msg.mentions.members.first();
    var reason = args;
    var silentMsg = "";
    if (!user) {
        msg.reply("Musisz podać użytkownika! `!ban <id>`");
        return
    }
    if (!user.kickable) {
        msg.reply("Nie mogę zbanować tego użytkownika!")
        return
    }
    if (user == msg.author) {
        msg.reply("Dlaczego chcesz zbanować samego siebie?")
        return
    }
    if (!reason) {
        // eslint-disable-next-line no-redeclare
        var reason = "Brak powodu"
    }
    if (msg.content.includes("-s")) {
        var silent = true;
        // eslint-disable-next-line no-redeclare
        var silentMsg = " [Flaga -s]"
    }
    msg.reply(`Zrozumiano, użytkownik został zbanowany.${silentMsg}`);
    if (!silent) user.send(`**${msg.guild.name}**: Zostałeś zbanowany za \`${reason}\``)
    msg.guild.members.ban(user, {
        reason: reason,
    });
}

var altban = function (msg) {
    
    if (msg.member.id == "337509297321803786" || msg.member.permissions.has("BAN_MEMBERS")) {
        var args = msg.content.trim().split(' ');
        var user = bot.client.users.cache.find(user => user.id === args[0])
        //bot.users.cache.find(user => user.id === 'USER-ID')
        //args.split(' ').splice(0, 1);
        //args.splice(0, 1);
        if (!user) {
            msg.reply("Musisz podać użytkownika! `!altban <id>`");
            return
        }
        if (!user.kickable) {
            msg.reply("Nie mogę zbanować tego użytkownika!")
            return
        }
        if (user == msg.author) {
            msg.reply("Dlaczego chcesz zbanować samego siebie?")
            return
        }
        msg.reply(`Zrozumiano, użytkownik został zbanowany.`);
        //if (!silent) user.send(`**${msg.guild.name}**: Zostałeś zbanowany za \`Alt konto\``)
        msg.guild.members.ban(user, {
            reason: "Alt konto",
        });
    }
    else {
        msg.reply("Brak uprawnień!");
        return
    }
}

var setup = function (b) {
    bot = b;
    bot.registerCommand("ban", ban);
    bot.registerCommand("altban", altban);
    bot.registerCommand("oldban", banold);
}

exports.requires = [];
exports.setup = setup;

File: cogs/moderation/bans.py
import tomli, discord, re, json
from discord.ext import commands
from discord.ext.commands.core import has_permissions
from discord.ext.commands import MemberConverter

with open("configs/config.toml", "rb") as config:
    data = tomli.load(config)
    prefix = data["bot"]["prefix"]
    footer = data["modules"]["embeds"]["footerCopyright"]
    footer_img = data["modules"]["embeds"]["footerCopyrightImage"]
    administrators = data["bot"]["roles"]["administrators"]
    moderators = data["bot"]["roles"]["moderators"]

class Bans(commands.Cog):
    def __init__(self, bot, intents):
        self.bot = bot

    intents = discord.Intents.all()
    bot = commands.Bot(command_prefix=prefix, intents=intents)
    staff = administrators + moderators

    @bot.command(aliases=["zbanuj"])
    @commands.check_any(commands.has_any_role(*staff), commands.has_guild_permissions(ban_members=True))
    async def ban(self, ctx, member, *, reason = None):
        if ctx.message.mention_everyone:
            await ctx.reply("Na co liczysz?", mention_author=False, delete_after=15.0)
            return

        if len(ctx.message.mentions) > 1:
            await ctx.reply(f"Użyj komendy {prefix}multiban jeśli potrzebujesz zbanować kilku użytkowników.", mention_author=False, delete_after=15.0)
            return
        elif len(ctx.message.mentions) == 0:
            converter = MemberConverter()
            member = await converter.convert(ctx, member)
            if member is None:
                await ctx.reply("Nie znaleziono użytkownika.", mention_author=False, delete_after=15.0)
                return
        else:
            member = ctx.message.mentions[0]

        if member.id == ctx.author.id:
            await ctx.send("Dlaczego chcesz zbanować samego siebie?")
            return
        if member.id == self.bot.user.id:
            await ctx.send("Nie mogę zbanować samego siebie.")
            return
        days=0
        daysMsg = ""
        if "-24h" in ctx.message.content:
            days=1
            reason = reason.replace('-24h', '')
            daysMsg = "\nDodatkowo usunięto wiadomości z ostatnich 24 godzin."
        if "-7d" in ctx.message.content:
            days=7
            reason = reason.replace('-7d', '')
            daysMsg = "\nDodatkowo usunięto wiadomości z ostatnich 7 dni."

        silentMsg = "[Flaga -s]"
        if not "-s" in ctx.message.content:
            silentMsg = ""
            description = f"""Otrzymano banicję. Powód: {reason}"""
            embed=discord.Embed(description=description, color=0xff0000, timestamp=ctx.message.created_at)
            embed.set_author(name=ctx.guild.name)
            embed.set_footer(text=footer, icon_url=footer_img)
            try:
                await member.send(embed=embed)
            except:
                pass

        reason = reason.replace('-s', '')
        description = f"""**Nadano banicję.**\n
        **Użytkownik**: {member.mention} ({member.name}#{member.discriminator})
        **Administrator**: {ctx.author.mention} ({ctx.author.name}#{ctx.author.discriminator})
        **Powód**: {reason} {silentMsg} \n{daysMsg}"""
        embed2=discord.Embed(description=description, color=0xff0000, timestamp=ctx.message.created_at)
        embed2.set_author(name=ctx.guild.name)
        embed2.set_footer(text=footer, icon_url=footer_img)
        try:
            await member.ban(delete_message_days=days, reason = f"{reason} {silentMsg} {daysMsg}")
        except Exception as e:
            await ctx.send(f"Wystąpił błąd. [Bans-53] ```\n{e}\n```")
        await ctx.send(embed=embed2)


    @ban.error
    async def ban_error(self, ctx, error):
        if isinstance(error, commands.errors.CommandInvokeError):
            error = error.original
        if isinstance(error, commands.errors.MissingRequiredArgument):
            await ctx.reply(f"Nie podano wszystkich argumentów! \n```{prefix}ban <@Użytkowni(-k/-cy)/ID> <powód> (-s) (-24h/-7d) \n```", mention_author=False)
        elif isinstance(error, commands.errors.MissingPermissions):
            await ctx.reply("Brak uprawnień!", mention_author=False)
            raise error
        else:
            await ctx.send(f"Wystąpił błąd! **Treść**: \n```{error}```")

    @bot.command(aliases=["mban"])
    @commands.check_any(commands.has_any_role(*staff), commands.has_guild_permissions(ban_members=True))
    async def multiban(self, ctx, *, args):
        reason = None
        failed = []
        if ctx.message.mention_everyone:
            await ctx.reply("Na co liczysz?", mention_author=False, delete_after=15.0)
            return

        if len(ctx.message.mentions) == 0:
            ids = []
            reason = []
            members = []
            

            ids = re.findall("[0-9]{17,19}", args)
            reason = [x for x in args.split(" ") if x not in ids]
            reason = " ".join(reason)

            for id in ids:
                if str(id) == str(ctx.author.id):
                    failed.append((id, "Nie możesz zbanować samego siebie"))
                    continue
                elif str(id) == str(self.bot.user.id):
                    failed.append((id, "Podano ID bota"))
                    continue

                converter = MemberConverter()
                member = await converter.convert(ctx, member)
                bot_member = ctx.guild.get_member(self.bot.user.id)
                if member != None:
                    if member.top_role >= ctx.author.top_role:
                        failed.append((id, "Nie możesz zbanować użytkownika o identycznej hierarchii"))
                        continue
                    elif member.top_role >= bot_member.top_role:
                        failed.append((id, "Bot jest niższy hierarchią"))
                        continue
                    else:
                        members.append(member)
                else:
                    failed.append((id, "Błędne ID/Zbanowane"))
        else:
            members = ctx.message.mentions

        days=0
        daysMsg = ""
        if "-24h" in ctx.message.content:
            days=1
            reason = reason.replace('-24h', '')
            daysMsg = "\nDodatkowo usunięto wiadomości z ostatnich 24 godzin."
        if "-7d" in ctx.message.content:
            days=7
            reason = reason.replace('-7d', '')
            daysMsg = "\nDodatkowo usunięto wiadomości z ostatnich 7 dni."

        silentMsg = "[Flaga -s]"
        if "-s" not in ctx.message.content:
            silentMsg = ""
        else:
            reason = reason.replace('-s', '')

        for member in members:
            try:
                if ("-s" not in ctx.message.content) and (member.bot != True):
                    description = f"""Otrzymano banicję. Powód: {reason}"""
                    dm_embed=discord.Embed(description=description, color=0xff0000, timestamp=ctx.message.created_at)
                    dm_embed.set_author(name=ctx.guild.name)
                    dm_embed.set_footer(text=footer, icon_url=footer_img)
                    try:
                        await member.send(embed=dm_embed)
                    except:
                        pass
                await member.ban(delete_message_days=days, reason = f"{reason} {silentMsg} {daysMsg}")
            except discord.HTTPException as e:
                if e.code == 50013:
                    failed.append((member.id, "Nie można zbanować tego użytkownika, brak uprawnień."))
                continue
            except Exception as e:
                await ctx.send("Wystąpił błąd. [Bans-155]")
                print(e)
                raise

        members_formatted = []
        for member in members:
            members_formatted.append(f"<@{member.id}> ({member.name}#{member.discriminator})")
        members_formatted = "\n - ".join(members_formatted)

        description = f"""**Nadano banicje.**\n
        **Użytkownicy**: \n - {members_formatted}
        **Administrator**: <@{ctx.author.id}> ({ctx.author.name}#{ctx.author.discriminator})
        **Powód**: \n```{reason} {silentMsg} \n{daysMsg}```"""
        if failed != []:
            if members == []:
                description = ""
            description += "\n**Nieudane:**"
            for fail in failed:
                description += f"\n<@{fail[0]}> - {fail[1]}"
        reply_embed=discord.Embed(description=description, color=0xff0000, timestamp=ctx.message.created_at)
        reply_embed.set_author(name=ctx.guild.name)
        reply_embed.set_footer(text=footer, icon_url=footer_img)
        await ctx.send(embed=reply_embed)


    @multiban.error
    async def multiban_error(self, ctx, error):
        if isinstance(error, commands.errors.CommandInvokeError):
            error = error.original
        if isinstance(error, commands.errors.MissingRequiredArgument):
            await ctx.reply(f"Nie podano wszystkich argumentów! \n```{prefix}multiban <@Użytkownicy>/ID <powód> (-s) (-24h/-7d) \n```", mention_author=False)
        elif isinstance(error, commands.errors.MissingPermissions):
            await ctx.reply("Brak uprawnień!", mention_author=False)
            raise error
        else:
            await ctx.send(f"Wystąpił błąd! **Treść**: \n```{error}```")

    @bot.command()
    @commands.check_any(commands.has_any_role(*staff), commands.has_guild_permissions(ban_members=True))
    async def getbans(self, ctx, *, args):
        try:
            server_id = int(args)
            guild = self.bot.get_guild(server_id)

            if guild is None:
                await ctx.send("Guild not found. Please provide a valid server ID.")
                return

            # bans = await guild.bans(limit=2000)
            bans = [entry async for entry in guild.bans(limit=2000)]
            ban_list = []
            for ban_entry in bans:
                ban_list.append({
                    "user_id": ban_entry.user.id,
                    "username": str(ban_entry.user),
                    "reason": ban_entry.reason,
                })

            # Save to a JSON file
            file_name = f"bans_{server_id}.json"
            with open(file_name, "w") as json_file:
                json.dump(ban_list, json_file, indent=4)

            await ctx.reply(content=f"Wyeksportowano {len(ban_list)} banów", file=discord.File(file_name))

        except ValueError:
            await ctx.send("Please provide a valid numeric server ID.")

        except discord.Forbidden:
            await ctx.send("I don't have permission to fetch bans for this server.")

        except Exception as e:
            await ctx.send(f"An error occurred: {e}")

async def setup(bot):
    intents = discord.Intents.default()
    intents.members = True
    await bot.add_cog(Bans(bot, intents))